Explain event callbacks in "Monitoring for parameter changes"

danielcranston opened the pull request:
This PR has some change suggestions that came up in https://github.com/ros2/kilted_tutorial_party/issues/357#issuecomment-2848922147.

The main part of the PR is moving the already-existing explanation on monitoring parameters from remote nodes to a higher level section (so it's visible in the ToC), and **adding a new section** that explains [add_parameter_event_callback](https://github.com/ros2/rclpy/blob/kilted/rclpy/rclpy/parameter_event_handler.py#L147).

Here's a repo/commit that applies the steps that this new subsection explains, in case it's helpful for reviewers: https://github.com/danielcranston/ros2_testing_repo/commit/aff00eab9a3fa41ae967eaef0a040eaa246e8c8d

---

~So far I only updated the Python version of this document. If this gets positive feedback I can give the C++ version a similar treatment as well.~ C++ version added as well.

danielcranston committed:
fixup some small nits

danielcranston committed:
lift 'remote node' one level higher, add event callback explanation

It currently exists as a subsection of "3 Build and run", which makes it
somewhat hidden. It also makes it hidden from the ToC.

This commit lifts it one level higher, as a subsection to a new
"Extensions" section.

Further, a second subsection to "Extensions" is added that explains that
event callbacks exists, and how to use them.

danielcranston commented:
Actually, I realize now that I don't have a good grasp of the difference between
* `ParameterEventHandler.add_parameter_event_callback`, and something like
* `Node.add_on_set_parameters_callback`

Feels like they cater to pretty much the same use case. Is their difference explained somewhere?

fujitatomoya commented on source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-Python.rst:
```diff
@@ -87,7 +87,6 @@ Inside the ``ros2_ws/src/python_parameter_event_handler/python_parameter_event_h
 
             self.callback_handle = self.handler.add_parameter_callback(
                 parameter_name="an_int_param",
-                node_name="node_with_parameters",
```
> In the original pasted Python code, specifying node_name="node_with_parameters" to self.handler.add_parameter_callback is redundant.

can you elaborate a bit this? i do not see why this is redundant...

https://github.com/ros2/rclpy/blob/3414456ddaf1163d1951d301336028c9600bd58e/rclpy/rclpy/parameter_event_handler.py#L292

it looks like required argument here to specify this test node's parameter event activity.

danielcranston replied:
Oh true, thanks for catching this! I was looking at the rclcpp version, where it is optional.

https://github.com/ros2/rclcpp/blob/4b1c125/rclcpp/include/rclcpp/parameter_event_handler.hpp#L235-L252

I'll revert, but maybe as a separate issue it'd be nice to make the API identical between rclpy and rclcpp.

fujitatomoya commented on source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-Python.rst:
```diff
@@ -262,8 +260,16 @@ The terminal running the node will display a message similar to the following:
 
-3.1 Monitor changes to another node's parameters
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Extensions
+----------
+
+So far, we built and tested a small node that monitors a single parameter owned by the node itself.
+Using this node as a base, two other usecases where the ParameterEventHandler can be useful is presented below.
+
+
+Monitor changes to another node's parameters
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
i think original doc explains well, especially we can find the following description. i would keep here as it is, since i do not really see the difference.

```rst
You can also use the ParameterEventHandler to monitor parameter changes to another node's parameters.
Let's update the SampleNodeWithParameters class to monitor for changes to a parameter in another node.
```

danielcranston replied:
Thanks for the review! Just to check my understanding, you're saying to revert this block of additions (Your text in code block appears just below these lines), right? 

I realize I didn't make the intent very clear (what does ["lift 'remote node'  one level higher"](https://github.com/ros2/ros2_documentation/pull/5543/commits/5c57c8ae67c7e7a38361c3c4e709831bde101e6c) mean?). This aims to make the "Tasks" section only talk about the basic usage of `add_parameter_callback`, and delegate further discussion to a separate "Extensions" section (at the same level as "Tasks").

So we'd go from:

> * Background
> * Prerequisites
> * Tasks
>   - 1 Create a package
>   - 2 Write the Python node
>   - 3 Build and run **<-- subsections (3.1, 3.2) are hidden from the ToC**
> * Summary
> * Related content

To:


> * Background
> * Prerequisites
> * Tasks
>   - 1 Create a package
>   - 2 Write the Python node
>   - 3 Build and run <-- no more subsections, it only talks about the basic usage
> * **Extensions**
>   - **Monitor changes to another node's parameters**
>   - **Monitor all node parameters simultaneously**
> * Summary
> * Related content

If this was already understood and you still prefer to revert I'm fine, just making sure. Personally I think having these visible in the ToC is nicer.

It's also a bit surprising for "_Monitor changes to another node's parameters_" and "_Monitor all node parameters simultaneously_" to be subsections of something called "_Build and run_"

fujitatomoya replied:
I see, thanks for your explanation. i think your proposed sections are better, i will go ahead to resolve the related comments.

fujitatomoya commented on source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-Python.rst:
```diff
@@ -334,11 +340,115 @@ Upon executing this command, you should see output in the parameter_event_handle
       [INFO] [1699821958.757770223] [node_with_parameters]: Received an update to parameter: a_double_param: 3.45
 
+
+Monitor all node parameters simultaneously
```
this is great enhancement 🚀 to explain user how to use `add_parameter_event_callback`.
can we sectionize this?

```suggestion
3.2 Monitor all node parameters simultaneously
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

danielcranston replied:
See the previous comment.

fujitatomoya commented on source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-Python.rst:
```diff
@@ -334,11 +340,115 @@ Upon executing this command, you should see output in the parameter_event_handle
+Monitor all node parameters simultaneously
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If you have a node with lots of parameters, it would be cumbersome to have to call ``add_parameter_callback`` once for each parameter.
```
```suggestion
If you need to monitor multiple nodes or parameters at the same time, it would be cumbersome to have to call ``add_parameter_callback`` once for each of them.
```

danielcranston replied:
I might be wrong, but it looks to me that `add_parameter_event_callback` can only be used to monitor self-owned parameters (in Python at least), not remote ones. Can you confirm?

fujitatomoya replied:
i do not think that limitation, it just fetches the event data from `/parameter_events` topic, and issue the callback. (rclcpp has the consistent behavior with rclpy, as far as i see.)

danielcranston replied:
Ah yes, I should have looked at the implementation! Thanks for the suggestion.

danielcranston committed:
revert removal of node_name arg in add_parameter_callback call

danielcranston commented on source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-Python.rst:
```diff
@@ -334,11 +342,115 @@ Upon executing this command, you should see output in the parameter_event_handle
+      [INFO] [1746414766.243499816] [node_with_parameters]: Received parameter event
+      [INFO] [1746414766.244271445] [node_with_parameters]: Inside event: an_int_param changed to: 4
+
+Now set the new doble parameter:
```
Oops. Should we add `doble->double` to https://github.com/ros2/ros2_documentation/blob/rolling/codespell_dictionary.txt ? :smile: 
```suggestion
Now set the new double parameter:
```

danielcranston committed:
Fix typo

Signed-off-by: danielcranston <45213235+danielcranston@users.noreply.github.com>

fujitatomoya commented:
@danielcranston almost lgtm, can you check my last comment?

danielcranston committed:
Update source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-Python.rst

Co-authored-by: Tomoya Fujita <Tomoya.Fujita@sony.com>
Signed-off-by: danielcranston <45213235+danielcranston@users.noreply.github.com>

danielcranston committed:
update cpp version equivalently

danielcranston commented:
Thanks for the review @fujitatomoya, I addressed the outstanding comment, and updated the C++ doc as well in https://github.com/ros2/ros2_documentation/pull/5543/commits/af2afbcf17863ebc3377332a30f3fd103d10afdf.

fujitatomoya commented:
@christophebedard i would like to have 2nd review for this, can you take a look when you get back?

christophebedard commented on source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-CPP.rst:
```diff
@@ -250,8 +250,16 @@ The terminal running the node will display a message similar to the following:
+----------
+
+So far, we built and tested a small node that monitors a single parameter owned by the node itself.
+Using this node as a base, two other usecases where the ParameterEventHandler can be useful is presented below.
```
```suggestion
Using this node as a base, two other usecases where the ParameterEventHandler can be useful are presented below.
```

christophebedard commented on source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-Python.rst:
```diff
@@ -262,8 +262,16 @@ The terminal running the node will display a message similar to the following:
+----------
+
+So far, we built and tested a small node that monitors a single parameter owned by the node itself.
+Using this node as a base, two other usecases where the ParameterEventHandler can be useful is presented below.
```
```suggestion
Using this node as a base, two other usecases where the ParameterEventHandler can be useful are presented below.
```

christophebedard commented on source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-CPP.rst:
```diff
@@ -250,8 +250,16 @@ The terminal running the node will display a message similar to the following:
 
-3.1 Monitor changes to another node's parameters
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
```
Nitpick: only 1 blank line between sections/anything, like before.

```suggestion
```

christophebedard commented on source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-CPP.rst:
```diff
@@ -250,8 +250,16 @@ The terminal running the node will display a message similar to the following:
+So far, we built and tested a small node that monitors a single parameter owned by the node itself.
+Using this node as a base, two other usecases where the ParameterEventHandler can be useful is presented below.
+
+
```
Nitpick: keep 1 blank line between sections, like before.

```suggestion
```

christophebedard commented on source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-CPP.rst:
```diff
@@ -338,11 +346,122 @@ Upon executing this command, you should see output in the parameter_event_handle
 
     [INFO] [1606952588.237531933] [node_with_parameters]: cb2: Received an update to parameter "a_double_param" of type: double: "3.45"
 
+
```
Nitpick: keep 1 blank line between sections, like before.

```suggestion
```

christophebedard commented on source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-CPP.rst:
```diff
@@ -338,11 +346,122 @@ Upon executing this command, you should see output in the parameter_event_handle
+        };
+      };
+    event_cb_handle_ = param_subscriber_->add_parameter_event_callback(event_cb);
+
```
```suggestion
```

christophebedard commented on source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-CPP.rst:
```diff
@@ -338,11 +346,122 @@ Upon executing this command, you should see output in the parameter_event_handle
+    private:
+      ...
+      std::shared_ptr<rclcpp::ParameterEventCallbackHandle> event_cb_handle_;
+
```
```suggestion
```

christophebedard commented on source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-CPP.rst:
```diff
@@ -338,11 +346,122 @@ Upon executing this command, you should see output in the parameter_event_handle
+    .. code-block:: console
+
+      $ call install\setup.bat
+
```
```suggestion
```

christophebedard commented on source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-CPP.rst:
```diff
@@ -338,11 +346,122 @@ Upon executing this command, you should see output in the parameter_event_handle
+      [INFO] [1747144403.418980063] [node_with_parameters]: cb: Received an update to parameter "an_int_param" of type integer: "44"
+      [INFO] [1747144403.419086611] [node_with_parameters]: Received parameter event
+      [INFO] [1747144403.419114103] [node_with_parameters]: Inside event: "an_int_param" changed to 44
+
```
```suggestion
```

christophebedard commented on source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-CPP.rst:
```diff
@@ -338,11 +346,122 @@ Upon executing this command, you should see output in the parameter_event_handle
+   When setting multiple parameters at once, it's best to use ``set_parameters_atomically``, explained in :doc:`../../Concepts/Basic/About-Parameters`.
+   This way, the event callback is only fired once.
+
+
```
```suggestion
```

christophebedard commented on source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-Python.rst:
```diff
@@ -262,8 +262,16 @@ The terminal running the node will display a message similar to the following:
 
-3.1 Monitor changes to another node's parameters
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
```
Nitpick: keep 1 blank line between sections, like before.

```suggestion
```

christophebedard commented on source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-Python.rst:
```diff
@@ -262,8 +262,16 @@ The terminal running the node will display a message similar to the following:
+So far, we built and tested a small node that monitors a single parameter owned by the node itself.
+Using this node as a base, two other usecases where the ParameterEventHandler can be useful is presented below.
+
+
```
Nitpick: keep 1 blank line between sections, like before.

```suggestion
```

christophebedard commented on source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-Python.rst:
```diff
@@ -334,11 +342,115 @@ Upon executing this command, you should see output in the parameter_event_handle
 
       [INFO] [1699821958.757770223] [node_with_parameters]: Received an update to parameter: a_double_param: 3.45
 
+
```
Nitpick: keep 1 blank line between sections, like before.

```suggestion
```

christophebedard commented on source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-Python.rst:
```diff
@@ -334,11 +342,115 @@ Upon executing this command, you should see output in the parameter_event_handle
+            self.get_logger().info(
+                f"Inside event: {p.name} changed to: {rclpy.parameter.parameter_value_to_python(p.value)}"
+            )
+
```
```suggestion
```

christophebedard commented on source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-Python.rst:
```diff
@@ -334,11 +342,115 @@ Upon executing this command, you should see output in the parameter_event_handle
+
+Note that the ``parameter_event`` is of type `rcl_interfaces/ParameterEvent <https://docs.ros.org/en/rolling/p/rcl_interfaces/msg/ParameterEvent.html>`_.
+Although it's not shown in this tutorial, event callbacks can also be used to monitor when parameters are added or deleted.
+
```
```suggestion
```

christophebedard commented on source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-Python.rst:
```diff
@@ -334,11 +342,115 @@ Upon executing this command, you should see output in the parameter_event_handle
+    .. code-block:: console
+
+      $ call install\setup.bat
+
```
```suggestion
```

christophebedard commented on source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-Python.rst:
```diff
@@ -334,11 +342,115 @@ Upon executing this command, you should see output in the parameter_event_handle
+
+   When setting multiple parameters at once, it's best to use ``set_parameters_atomically``, explained in :doc:`../../Concepts/Basic/About-Parameters`.
+   This way, the event callback is only fired once.
+
```
```suggestion
```

christophebedard commented on source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-Python.rst:
```diff
@@ -334,11 +342,115 @@ Upon executing this command, you should see output in the parameter_event_handle
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If you need to monitor multiple nodes or parameters at the same time, it would be cumbersome to have to call ``add_parameter_callback`` once for each of them.
+In this case, you can use ``add_parameter_event_callback`` to register a single callback that fires when *any* of the node's parameters change.
```
Does `add_parameter_event_callback()` only trigger the given callback when a specific node's parameters change, or does it trigger it when _any_ parameters of _any_ nodes change?

danielcranston replied:
The latter. I pushed 9d63c3840bc8ecbaa9e95b97b74fad31738dad17 to make this more explicit.

christophebedard commented on source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-CPP.rst:
```diff
@@ -338,11 +346,122 @@ Upon executing this command, you should see output in the parameter_event_handle
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If you need to monitor multiple nodes or parameters at the same time, it would be cumbersome to have to call ``add_parameter_callback`` once for each of them.
+In this case, you can use ``add_parameter_event_callback`` to register a single callback that fires when *any* of the node's parameters change.
```
Same question here: is this only for a specific node's parameters, or is it for all parameters of all nodes?

christophebedard commented on source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-CPP.rst:
```diff
@@ -338,11 +346,122 @@ Upon executing this command, you should see output in the parameter_event_handle
+
+
+This declares a new double parameter ``another_double_param`` and adds an event callback that will monitor both parameters.
+Note that the ``parameter_event`` is of type `rcl_interfaces/ParameterEvent <https://docs.ros.org/en/rolling/p/rcl_interfaces/msg/ParameterEvent.html>`_.
```
```suggestion
Note that the ``parameter_event`` is of type {interface(rcl_interfaces/msg/ParameterEvent)}.
```

christophebedard commented on source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-Python.rst:
```diff
@@ -334,11 +342,115 @@ Upon executing this command, you should see output in the parameter_event_handle
+
+     $ ros2 run python_parameter_event_handler node_with_parameters
+
+Then, from a second terminal (with ROS initialized), let's set the original int parameter:
```
I'm assuming you mean "sourcing," so it would be better to say that, since that's what we call it everywhere:

```suggestion
Then, from a second terminal (with ROS sourced), let's set the original int parameter:
```

christophebedard commented on source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-CPP.rst:
```diff
@@ -338,11 +346,122 @@ Upon executing this command, you should see output in the parameter_event_handle
+
+     $ ros2 run cpp_parameter_event_handler parameter_event_handler
+
+Then, from a second terminal (with ROS initialized), let's set the original int parameter:
```
I'm assuming you mean "sourcing," so it would be better to say that, since that's what we call it everywhere:

```suggestion
Then, from a second terminal (with ROS sourced), let's set the original int parameter:
```

christophebedard commented on source/Tutorials/Intermediate/Monitoring-For-Parameter-Changes-Python.rst:
```diff
@@ -334,11 +342,115 @@ Upon executing this command, you should see output in the parameter_event_handle
+            )
+
+
+Note that the ``parameter_event`` is of type `rcl_interfaces/ParameterEvent <https://docs.ros.org/en/rolling/p/rcl_interfaces/msg/ParameterEvent.html>`_.
```
```suggestion
Note that the ``parameter_event`` is of type {interface(rcl_interfaces/msg/ParameterEvent)}.
```

danielcranston committed:
Apply suggestions from code review

Co-authored-by: Christophe Bedard <bedard.christophe@gmail.com>
Signed-off-by: danielcranston <45213235+danielcranston@users.noreply.github.com>

danielcranston committed:
make explicit that event callbacks trigger for any param in any node

christophebedard reviewed (approved):
Looks good, thanks for making the changes!
